lib/api.js

```diff
@@ -0,0 +1,113 @@
+import base from 'taskcluster-base';
+import common from './common';
+import storage from './storage';
+import _ from 'lodash';
+
+/**
+ * Check to see if a resource an expiration has passed....
+ **/
+function hasExpired(expires) {
+  let currentTime = new Date();
+  let expireTime = new Date(expires);
+  if (currentTime > expireTime) {
+    return true;
+  }
+  return false;
+};
+
+/** API end-point for version v1/
+ *
+ * In this API implementation we shall assume the following context:
+ * {
+ *   publisher:      // publisher from base.Exchanges
+ * }
+ */
+var api = new base.API({
+  title:        "TaskCluster Secrets API Documentation",
+  description: [
+    "The secrets service, typically available at",
+    "`tools.taskcluster.net`, is responsible for managing",
+    "secure data via TaskCluster scopes."
+  ].join('\n')
+});
+
+// Export API
+module.exports = api;
+
+/** Define tasks */
+api.declare({
+  method:      'put',
+  route:       '/scope/:secretScope',
+  name:        'secretWrite',
+  input:       common.SCHEMA_PREFIX_CONST + "secret.json#",
+  scopes:      ['secrets:write'],
+  title:       'Create New Secret',
+  description: 'Set a secret associated with some scope.'
+}, async function(req, res) {
+    let {secretScope} = req.params;
+    let params = req.body;
+    req.satisfies([secretScope])
+    if (params.expires) {
+      let expires = new Date(params.expires);
+      if (expires == 'Invalid Date') {
+        res.status(400).send('expires: must be a datetime (YYYY-MM-DD hh:mm:ss)');
+        return;
+      }
+    }
+    this.storageObject.write(secretScope, params);
```

Comment:
When using azure table storage it won't be this simple. You have to ensure the operation is idempotent.

This is typically done as follows:

```js
try{
  this.Secret.create({
    name: ...,
    secret: ...,
    expires: ...
}.catch(err) {
  if (err.code !== 'EntityAlreadyExists') {
    throw err;
  }
  
  var secret = await this.Secrets.load({name: ...});
  if (secret.expires.getTime() !== new Date(req.body.expires).getTime() ||
      !_.isEqual(secrets.secret, req.body.secret)) {
    return res.status(409).json({message: "Secret already exists!"});
  }
}
return res.status(201).send();
```
For an example see: https://github.com/taskcluster/taskcluster-queue/blob/master/routes/v1.js#L298-L344

The idea is that if the entity already exists... and matches the one you're trying to create... Then you return success, as the operation is idempotent. It's probably a retry of the request... This is smart :)